### ISSUE TITLE
slackSend broken in 2.18

### ISSUE THREAD

**Issue by brenc** [jenkinsci/slack-plugin#539]
Upgraded from 2.14 to 2.18 recently which appears to have broken `slackSend`. No related logs in the console nor in jenkins.log. Here is an example command we're using:

`slackSend channel: '#brens_projects', color: '#FFFF00', message: "STARTED: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]' (${env.BUILD_URL})", teamDomain: 'foobar', token: 'the-token'`

I reverted back to 2.14 and it's working again.

Edit: Devuan Linux 2 (Debian 9/Stretch), Jenkins 2.168, Java 1.8.0_181.

**Comment by reilly3000** [comment]
I'm having the same issue with 2.18. I just installed it so I'm not really sure how to rollback.

**Comment by brenc** [comment]
Managing Jenkins -> Manage Plugins. Click Installed tab. Search for Slack Notification. Click the downgrade button. Click the restart Jenkins box.

**Comment by reilly3000** [comment]
Unfortunately I don't have a downgrade link. Is there any way I can install 2.14 manually? I don't see any releases for that I can point to in this repo.

FYI the plugin is working fine for me with a freestyle project, just not from my Jenkinsfile. Thanks for your help!

**Comment by timja** [comment]
Can you:
1. enable fine logs for StandardSlackService (https://github.com/jenkinsci/slack-plugin/blob/master/src/main/java/jenkins/plugins/slack/StandardSlackService.java)
2. try a credential instead of a token

**Comment by reilly3000** [comment]
I just had a successful Slack message post from my pipeline using v2.18, the logs revealed a syntax error with my message.

**Comment by brenc** [comment]
@timja - can you explain a bit more? I Googled around a bit but I'm not sure exactly how to enable fine logs.

This might be a separate bug but the plugin doesn't work using credentials. Never has. That's why I have it using a token within slackSend.

**Comment by timja** [comment]
The plugin works with credentials if they are set at the global level
there's a bug which will be worked on soon to fix that

https://support.cloudbees.com/hc/en-us/articles/204880580-How-do-I-create-a-logger-in-Jenkins-for-troubleshooting-and-diagnostic-information-

**Comment by brenc** [comment]
I set the credentials at the global level but could never get it to work. Once I added the credential in the slackSend command itself, it worked. I'll have to revisit this.

I upgraded to 2.18, enabled fine logs for StandardSlackService, and switched to using the global credential. I see no errors:

```Mar 15, 2019 1:22:23 PM FINE jenkins.plugins.slack.StandardSlackService
Posting succeeded```

The post parameters all looks fine. I think I'm going to downgrade to 2.14 and compare the post parameters with the output from 2.18.

**Comment by brenc** [comment]
OK so 2.18 posts to `https://slack.com/api/chat.postMessage` and 2.14 posts to `https://newgrounds.slack.com/services/hooks/jenkins-ci`. I think this is the problem.

**Comment by brenc** [comment]
Double checked the global settings and the global webhook URL endpoint IS set correctly, but 2.18 isn't using it for some reason.

**Comment by timja** [comment]
`chat.postMessage` means that you have `botUser` enabled somewhere.

**Comment by brenc** [comment]
Hrmm - where would that be set? I vaguely remember seeing it somewhere.

Here's what I've observed: using 2.14, if I use a command like this (no token):

`slackSend channel: '#brens_projects', color: '#FFFF00', message: "STARTED: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]' (${env.BUILD_URL})", teamDomain: 'newgrounds'`

It posts to `https://slack.com/api/chat.postMessage` and no message comes through. If I add the token parameter to it, it will post to `https://newgrounds.slack.com/services/hooks/jenkins-ci` and the message _does_ come through.

This is similar to 2.18 with the exception that it _always_ posts to `https://slack.com/api/chat.postMessage` and a message never comes through.

**Comment by timja** [comment]
It would be set globally,
it should be in the build log saying what the current value is

**Comment by brenc** [comment]
I don't have that set anywhere.

**Comment by timja** [comment]
What does it say in the build log?

**Comment by brenc** [comment]
Is this what you're asking for?

```
run slackstepsend, step false, desc true
Slack Send Pipeline step configured values from global config - baseUrl: true, teamDomain: false, token: false, channel: false, color: false
```

**Comment by timja** [comment]
Yes, can you unset baseUrl and use teamDomain instead please

**Comment by brenc** [comment]
I don't have baseUrl set anywhere, unless you mean the global webhook setting. I'm using teamDomain already. Again, here's the full command I'm using:

`slackSend channel: '#brens_projects', color: '#FFFF00', message: "STARTED: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]' (${env.BUILD_URL})", teamDomain: 'newgrounds', token: 'THE TOKEN'`

**Comment by timja** [comment]
can you screenshot your global config please?

**Comment by brenc** [comment]
It's been a while since I set this up so I totally forgot that there are two sections for Slack in global settings: **Slack Webhook Settings** and **Global Slack Notifier Settings**. I stumbled upon the latter section today.

Looks like I *did* have "Is bot user?* ticked. I probably figured that was the correct setting and frankly I'm still confused what the setting is for. So I unticked that box and redid the `slackSend` commands to this:

`slackSend channel: '#brens_projects', color: '#FFFF00', message: "STARTED: Job '${env.JOB_NAME} [${env.BUILD_NUMBER}]' (${env.BUILD_URL})"`

And now they are all working on 2.18.

For the record, under **Global Slack Notifier Settings** every field is blank / default except `Team Subdomain` and `Integration Token Credential ID` which I have set appropriately.

I'll close the ticket now. Thanks for the help.
